diff --git a/lib/libutee/include/tee_internal_api.h b/lib/libutee/include/tee_internal_api.h
index 6ebeaebea..5f5ad812b 100644
--- a/lib/libutee/include/tee_internal_api.h
+++ b/lib/libutee/include/tee_internal_api.h
@@ -17,6 +17,23 @@
 #include <tee_api_types.h>
 #include <trace.h>
 
+#define SHM_MITIGATION 1
+
+#if SHM_MITIGATION
+struct ShmInfo {
+    uint32_t    commandID;
+    bool        params2shared[4]; 
+    struct ShmInfo*    next;    
+};
+typedef struct ShmInfo ShmInfo_t;
+
+extern ShmInfo_t *Invoke_Shms;
+extern ShmInfo_t *Open_Shms;
+
+TEE_Result TEE_RegisterShm(uint32_t commandID, bool* params2shared, bool is_OpenSession); 
+TEE_Result TEE_IsShared(uint32_t commandID, uint32_t params_idx, bool is_OpenSession, bool* out);
+#endif 
+
 /* Property access functions */
 
 TEE_Result TEE_GetPropertyAsString(TEE_PropSetHandle propsetOrEnumerator,
diff --git a/lib/libutee/tee_api.c b/lib/libutee/tee_api.c
index 1ca98e236..923fabb43 100644
--- a/lib/libutee/tee_api.c
+++ b/lib/libutee/tee_api.c
@@ -31,6 +31,67 @@ struct malloc_ctx *__ta_no_share_malloc_ctx __weak;
 
 static const void *tee_api_instance_data;
 
+#if SHM_MITIGATION
+ShmInfo_t* Invoke_Shms = NULL;
+ShmInfo_t* Open_Shms = NULL;
+TEE_Result TEE_RegisterShm(uint32_t commandID, bool* params2shared, bool is_OpenSession){
+    if(is_OpenSession){
+        Open_Shms = (ShmInfo_t*)TEE_Malloc(sizeof(ShmInfo_t),0);
+        memset(Open_Shms, 0, sizeof(ShmInfo_t));
+        memcpy(&Open_Shms->params2shared, params2shared, 4*sizeof(bool));
+        Open_Shms->next = NULL;
+        Open_Shms->commandID = 0;
+        return TEE_SUCCESS;
+    }
+    ShmInfo_t* new_shm = (ShmInfo_t*)TEE_Malloc(sizeof(ShmInfo_t),0);
+    memset(new_shm, 0, sizeof(ShmInfo_t));
+    new_shm->commandID = commandID;
+    new_shm->next = NULL;
+    memcpy(&new_shm->params2shared, params2shared, 4*sizeof(bool));
+    if(Invoke_Shms == NULL){
+        EMSG("new InvokeShm structure is registered"); 
+        Invoke_Shms = new_shm;
+    } else {
+        ShmInfo_t* cur = Invoke_Shms;
+        while(cur->next != NULL){
+            cur = cur->next;
+        }
+        EMSG("inserting new command into linked list"); 
+        cur->next = new_shm;
+    }
+    return TEE_SUCCESS;
+}
+
+TEE_Result TEE_IsShared(uint32_t commandID, uint32_t params_idx, bool is_OpenSession, bool* out){
+    if(is_OpenSession){
+        if(Open_Shms == NULL){
+            *out = false;
+            return TEE_SUCCESS;
+        }
+        *out = Open_Shms->params2shared[params_idx];
+        return TEE_SUCCESS;
+    } 
+    if(Invoke_Shms == NULL){
+        EMSG("InvokeShms is NULL, returning default false!");
+        *out = false;
+        return TEE_SUCCESS;
+    } else {
+        ShmInfo_t* cur = Invoke_Shms;
+        do{
+            if(cur->commandID == commandID){
+                EMSG("matching commandID found, comparing paramTypes");
+                *out = cur->params2shared[params_idx]; 
+                return TEE_SUCCESS;
+            }
+            cur = cur->next;
+        } while(cur != NULL);
+        EMSG("commandID not found in registered shms, returning false!");
+        *out = false;
+        return TEE_SUCCESS;   
+    }   
+}
+#endif
+
 /* System API - Internal Client API */
 
 static void copy_param(struct utee_params *up, uint32_t param_types,
@@ -156,6 +217,7 @@ static TEE_Result map_tmp_param(struct utee_params *up, void **tmp_buf,
 		case TEE_PARAM_TYPE_MEMREF_INPUT:
 			if (!is_tmp_mem[n])
 				break;
+            EMSG("??????????????????????????????????????? allocating tmp buf");
 			s = up->vals[n * 2 + 1];
 			b = (void *)(vaddr_t)up->vals[n * 2];
 			tmp_va[n] = tb;
diff --git a/lib/libutee/user_ta_entry.c b/lib/libutee/user_ta_entry.c
index 13ff14c2b..272db4e1d 100644
--- a/lib/libutee/user_ta_entry.c
+++ b/lib/libutee/user_ta_entry.c
@@ -275,7 +275,7 @@ static void ta_header_remove_session(uint32_t session_id)
 }
 
 static void to_utee_params(struct utee_params *up, uint32_t param_types,
-			   const TEE_Param params[TEE_NUM_PARAMS])
+			   const TEE_Param params[TEE_NUM_PARAMS], void** tmpBufs, void** origBufs, uint32_t commandID, bool isOpenSession)
 {
 	size_t n = 0;
 
@@ -291,8 +291,29 @@ static void to_utee_params(struct utee_params *up, uint32_t param_types,
 		case TEE_PARAM_TYPE_MEMREF_INPUT:
 		case TEE_PARAM_TYPE_MEMREF_OUTPUT:
 		case TEE_PARAM_TYPE_MEMREF_INOUT:
-			up->vals[n * 2] = (uintptr_t)params[n].memref.buffer;
+#if SHM_MITIGATION
+            bool isShared = false;
+            TEE_IsShared(commandID, n, isOpenSession, &isShared);
+            if(isShared || (params[n].memref.buffer == 0 && params[n].memref.size == 0)){
+			    up->vals[n * 2] = (uintptr_t)params[n].memref.buffer;
+			    up->vals[n * 2 + 1] = params[n].memref.size;
+            } else {
+                EMSG("tmpBufs[n]= %p, origBufs[n]= %p", tmpBufs[n], origBufs[n]);
+                if(TEE_PARAM_TYPE_GET(param_types, n) != TEE_PARAM_TYPE_MEMREF_INPUT){
+                    memcpy(origBufs[n], params[n].memref.buffer, params[n].memref.size);
+                }
+                if(params[n].memref.size == 0x0){
+                    tee_unmap(tmpBufs[n], 0x1000);
+                } else {
+                    tee_unmap(tmpBufs[n], params[n].memref.size);
+                }
+			    up->vals[n * 2] = (uintptr_t)origBufs[n];
+			    up->vals[n * 2 + 1] = params[n].memref.size;
+            }
+#else
+            up->vals[n * 2] = (uintptr_t)params[n].memref.buffer;
 			up->vals[n * 2 + 1] = params[n].memref.size;
+#endif
 			break;
 		default:
 			up->vals[n * 2] = 0;
@@ -304,7 +325,8 @@ static void to_utee_params(struct utee_params *up, uint32_t param_types,
 
 static void from_utee_params(TEE_Param params[TEE_NUM_PARAMS],
 			     uint32_t *param_types,
-			     const struct utee_params *up)
+			     const struct utee_params *up,
+                 void** tmpBufs, void** origBufs, uint32_t commandID, bool isOpenSession)
 {
 	size_t n;
 	uint32_t types = up->types;
@@ -323,8 +345,30 @@ static void from_utee_params(TEE_Param params[TEE_NUM_PARAMS],
 		case TEE_PARAM_TYPE_MEMREF_INPUT:
 		case TEE_PARAM_TYPE_MEMREF_OUTPUT:
 		case TEE_PARAM_TYPE_MEMREF_INOUT:
-			params[n].memref.buffer = (void *)a;
-			params[n].memref.size = b;
+#if SHM_MITIGATION
+            bool isShared = false;
+            TEE_IsShared(commandID, n, isOpenSession, &isShared);
+            if(isShared || (a == 0 && b == 0)){
+                params[n].memref.buffer = (void *)a;
+                params[n].memref.size = b;
+            } else {
+                if(b == 0){
+                    tmpBufs[n] = tee_map_zi(0x1000, TEE_MEMORY_ACCESS_ANY_OWNER); //TEE_Malloc(b, 0);
+                } else {
+                    tmpBufs[n] = tee_map_zi(b, TEE_MEMORY_ACCESS_ANY_OWNER); //TEE_Malloc(b, 0);
+                }
+                origBufs[n] = (void*)a;
+                EMSG("size: %p, tmpBufs[n]: %p, origBufs[n]: %p", b, tmpBufs[n], origBufs[n]);
+                if(TEE_PARAM_TYPE_GET(types, n) != TEE_PARAM_TYPE_MEMREF_OUTPUT){
+                    memcpy(tmpBufs[n], (void*)a, b);
+                }
+                params[n].memref.buffer = tmpBufs[n];
+                params[n].memref.size = b;
+            }
+#else
+            params[n].memref.buffer = (void *)a;
+            params[n].memref.size = b;
+#endif
 			break;
 		default:
 			break;
@@ -342,6 +386,8 @@ static TEE_Result entry_open_session(unsigned long session_id,
 	struct ta_session *session;
 	uint32_t param_types;
 	TEE_Param params[TEE_NUM_PARAMS];
+    void* tmpBufs[4] = { NULL };
+    void* origBufs[4] = { NULL };
 
 	res = ta_header_add_session(session_id);
 	if (res != TEE_SUCCESS)
@@ -351,13 +397,13 @@ static TEE_Result entry_open_session(unsigned long session_id,
 	if (!session)
 		return TEE_ERROR_BAD_STATE;
 
-	from_utee_params(params, &param_types, up);
+	from_utee_params(params, &param_types, up, tmpBufs, origBufs, 0, true);
 	ta_header_save_params(param_types, params);
 
 	res = TA_OpenSessionEntryPoint(param_types, params,
 				       &session->session_ctx);
 
-	to_utee_params(up, param_types, params);
+	to_utee_params(up, param_types, params, tmpBufs, origBufs, 0, true);
 
 	if (res != TEE_SUCCESS)
 		ta_header_remove_session(session_id);
@@ -383,18 +429,21 @@ static TEE_Result entry_invoke_command(unsigned long session_id,
 	TEE_Result res;
 	uint32_t param_types;
 	TEE_Param params[TEE_NUM_PARAMS];
+    void* tmpBufs[4] =  { NULL };
+    void* origBufs[4] = { NULL };
+
 	struct ta_session *session = ta_header_get_session(session_id);
 
 	if (!session)
 		return TEE_ERROR_BAD_STATE;
 
-	from_utee_params(params, &param_types, up);
+	from_utee_params(params, &param_types, up, tmpBufs, origBufs, cmd_id, false);
 	ta_header_save_params(param_types, params);
 
 	res = TA_InvokeCommandEntryPoint(session->session_ctx, cmd_id,
 					 param_types, params);
 
-	to_utee_params(up, param_types, params);
+	to_utee_params(up, param_types, params, tmpBufs, origBufs, cmd_id, false);
 	return res;
 }
 
@@ -406,7 +455,7 @@ static TEE_Result entry_dump_memstats(unsigned long session_id __unused,
 	TEE_Param params[TEE_NUM_PARAMS] = { };
 	struct pta_stats_alloc stats = { };
 
-	from_utee_params(params, &param_types, up);
+	from_utee_params(params, &param_types, up, NULL, NULL, 0, false);
 	ta_header_save_params(param_types, params);
 
 	if (TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_OUTPUT,
@@ -422,7 +471,7 @@ static TEE_Result entry_dump_memstats(unsigned long session_id __unused,
 	params[1].value.b = stats.num_alloc_fail;
 	params[2].value.a = stats.biggest_alloc_fail;
 	params[2].value.b = stats.biggest_alloc_fail_used;
-	to_utee_params(up, param_types, params);
+	to_utee_params(up, param_types, params, NULL, NULL, 0, false);
 
 	return TEE_SUCCESS;
 }
